src/main/java/frc/robot/subsystems/Drivetrain.java

```diff
@@ -1,65 +1,525 @@
+  public SwerveModulePosition[] getModulePositions() {
+    SwerveModulePosition[] positions = new SwerveModulePosition[]{
+      m_modules[0].getPosition(),
+      m_modules[1].getPosition(),
+      m_modules[2].getPosition(),
+      m_modules[3].getPosition()
+    };
+    return positions;
+  }
+
+  /**
+   * Sets the desired states for all swerve modules.
+   * 
+   * @param swerveModuleStates an array of module states to set swerve modules to. Order of the array matters here!
+   */
+  public void setModuleStates(SwerveModuleState[] swerveModuleStates) {
+    for (int i = 0; i < 4; i++) {
+      m_modules[i].setDesiredState(swerveModuleStates[i]);
+    }
+  }
+
+  public void stop(){
+    for (int i = 0; i < 4; i++) {
+      m_modules[i].stop();
+    }
+  }
+
+  public void driveVoltsTest(double volts) {
+    // setAllOptimize(false);
+    for (int i = 0; i < 4; i++) {
+      m_modules[i].setDriveVoltage(volts);
+    }
+    m_modules[0].setSteerAngle(new Rotation2d(Units.degreesToRadians(135)));
+    m_modules[1].setSteerAngle(new Rotation2d(Units.degreesToRadians(45)));
+    m_modules[2].setSteerAngle(new Rotation2d(Units.degreesToRadians(225)));
+    m_modules[3].setSteerAngle(new Rotation2d(Units.degreesToRadians(315)));
+  }
+  public void steerVoltsTest(double volts) {
+    // setAllOptimize(false);
+    for (int i = 0; i < 4; i++) {
+      m_modules[i].setDriveVoltage(0);
+      m_modules[i].setSteerVoltage(volts);
+    }
+  }
+  public boolean isDriveVelocityAcurate(){
+    return 
+      Math.abs(m_modules[0].getDriveVelocityError()) < 0.1 &&
+      Math.abs(m_modules[1].getDriveVelocityError()) < 0.1 &&
+      Math.abs(m_modules[2].getDriveVelocityError()) < 0.1 &&
+      Math.abs(m_modules[3].getDriveVelocityError()) < 0.1;
+  }
+
+  public boolean isSteerAngleAcurate(){
+    return 
+      Math.abs(m_modules[0].getSteerAngleError()) < Units.degreesToRadians(1) &&
+      Math.abs(m_modules[1].getSteerAngleError()) < Units.degreesToRadians(1) &&
+      Math.abs(m_modules[2].getSteerAngleError()) < Units.degreesToRadians(1) &&
+      Math.abs(m_modules[3].getSteerAngleError()) < Units.degreesToRadians(1);
+  }
+
+  public double[] getDriveVelocities(){
+    return new double[] {
+      m_modules[0].getDriveVelocity(),
+      m_modules[1].getDriveVelocity(),
+      m_modules[2].getDriveVelocity(),
+      m_modules[3].getDriveVelocity()
+    };
+  }
+  public double[] getSteerVelocities(){
+    return new double[] {
+      m_modules[0].getSteerVelocity(),
+      m_modules[1].getSteerVelocity(),
+      m_modules[2].getSteerVelocity(),
+      m_modules[3].getSteerVelocity()
+    };
+  }
+  public void runCharacterizationVolts(int module, double value) {
+    for (int i = 0; i < 4; i++) {
+      m_modules[i].setDriveVoltage(0);
+      if (module == i){
+        m_modules[i].setSteerVoltage(value);
+      } else {
+        m_modules[i].setSteerVoltage(0);
+      }
+    }
+  }
+  
```

Comment:
Here and below: inconsistent formatting, should be only a single empty line after  function body closing bracket.

Reply:
In consistent sure, but there is a wider probolem of no clear standard line spaceing wich is something to discuss

Reply:
I think this is resolved but not sure as the line it is on has probably moved